### ISSUE TITLE
How are peaks from replicates handled in ChIP-seq and ATACseq pipelines?

### ISSUE THREAD

**Issue by AdrijaK** [maxplanck-ie/snakepipes#689]
Hi all, 

I have two questions about ATACseq and ChIPseq pipeline output. 
 
**1. How does snakePipes handle the peaks from replicates for CSAW analysis?**

_"The differential binding module utilizes the R package CSAW to detect significantly different peaks between two conditions. The analysis is performed on a union of peaks from all samples mentioned in the sample sheet."_ 

However, if I understand correctly, union of peaks from all replicates is used with MACS2, while Genrich produces union of replicate-concordant peaks for each treatment:

_"The output from Genrich will be peaks called per-group if you specify a sample sheet. This is because Genrich is capable of directly using replicates during peak calling."_

Is that correct? If yes, would it make sense to find replicate concordant peaks per group for MACS2 output using sampleSheet before merging them for CSAW?

**2. What is DiffBind peak score at column 5 in DiffBinding_allregions.bed file and DiffBinding_significant.bed?**

Which tool is this peak score derived from? is it diffBind -log10(pvalue)? I see that the range for DiffBinding_significant.bed starts from around 13 which would correspond to -10*log10(0.05). 

Thanks, 

Adrija

**Comment by katsikora** [comment]
Hi Adrija,

As to your question 1:

MACS2 is only able to output peaks per replicate (no replicate handling), while Genrich can handle replicates such that it outputs one peak file per "condition" group specified in the sample sheet.

In CSAW, the genome is segmented into sequential windows of defined size, in which read counts for each replicate are summarized. This returns a count matrix of genomic_bins x samples. This step is independent of peak calling.

In order to reduce the number of tests performed by CSAW, we intersect the genomic bins from the previous step with the union of peaks produced by the peak caller of choice. Further statistical analysis is performed on those genomic bins that intersected the peaks. There is no collapsing of replicates.  CSAW then performs an aggregation step over consecutive genomic windows to produce final list of differentially bound regions. Replicate group assignment is taken from the sample sheet.

Q2:

The score column in DiffBinding_allregions.bed is -log10(FDR) as returned by CSAW .
And "significant" is indeed filtered for the FDR threshold specified by the user, 0.05 is the default value.

Best,

Katarzyna

**Comment by AdrijaK** [comment]
Hi Katarzyna, 

Thank you for the detailed answer. 

>The score column in DiffBinding_allregions.bed is -log10(FDR) as returned by CSAW .

Does this mean that this score answers the question "which regions are differentially bound" and NOT "whether the peak is significantly enriched over the background"? 

Also, do you think introducing IDR-based selection of replicate-concordant peaks after MACS2 peak calling and before overlapping these peaks with CSAW bins would reduce the noise and increase the chances of using biologically relevant genomic regions for CSAW analysis? 

Best regards, 

Adrija

**Comment by katsikora** [comment]
Hi Adrija,

that's right, the pvalue and FDR from CSAW refer to the test on differentially bound regions between two experimental groups.
If you're looking for statistical evaluation of peak enrichment over background, you will find this information in the peak caller output files (MACS, Genrich, histoneHMM).

I'd not pre-select peaks for replicate concordance before evaluating them with CSAW - this would artificially lower the pvalues/FDR. It's the job of the statistical model in CSAW to use the information on log fold change between groups and within-group variance (i.e. the measure of variability between the replicates) to produce pvalues that reflect these .
In snakePipes, we only drop "empty" genomic bins from the initial count matrix by intersecting them with peaks - this is equivalent to filtering out lowly expressed genes in an RNA-seq count matrix to reduce the number of tests. This makes the pvalue correction for the number of tests to produce FDR estimates less stringent.
The replicate concordance is taken into consideration by the statistical model.

Hope this helps,

Best,

Katarzyna

**Comment by AdrijaK** [comment]
Thank you, this makes sense. 

Please correct me if I am wrong:
- DiffBinding_allregions.bed reports the union of peaks used  to filter CSAW bins. Then DiffBinding_significant.bed reports the PEAK coordinates overlapping the bins deemed as significantly regulated by CSAW. 
- If I want to look at peaks that are NOT differentially regulated (the peaks that stay constant between treatments), is it enough to choose regions in DiffBinding_allregions.bed that are not present in DiffBinding_significant.bed? 

Also:
- For ChIPseq the number of regions between DiffBinding_allregions.bed and Filtered.results.bed is the same. However, for ATACseq I see that the number of regions in DiffBinding_significant.bed is larger than in Filtered.results.bed. Is there additional ATAC-specific step that filters out certain peaks? 

```
wc -l  CSAW_MACS2_ATAC/*.bed
...
171225 CSAW_MACS2_ATAC/DiffBinding_allregions.bed
113484 CSAW_MACS2_ATAC/DiffBinding_significant.bed
106176 CSAW_MACS2_ATAC/Filtered.results.bed
...
```

- I also see that there has been a start coordinate shift by 1 between DiffBinding_significant.bed and Filtered.results.bed for both ATACseq and ChIPseq pipelines. Is intended? (SnakePipes 2.1.0)

```
==> CSAW_MACS2_ATAC/DiffBinding_significant.bed <==
chr1	3012700	3012780	region1	37.7837282285933	.

==> CSAW_MACS2_ATAC/Filtered.results.bed <==
chr1	3012701	3012780	80	*	37.7837282285933	4	4	0	5.08739260974428e-05	0.000166581656774165	up	3.27997328715966	3012761	region1
```

Thanks!

Best regards, 

Adrija

**Comment by katsikora** [comment]
- All the CSAW output bed files report genomic intervals based on *merged* genomic bins overlapping peaks.
This merging is done by CSAW functions `mergeWindows` and then `combineTests` . The idea is to merge adjacent genomic bins that have similar support for  differential binding into contiguous regions.
- The file DiffBinding_allregions.bed lists all merged regions output by CSAW without applying any filters.
- The file DiffBinding_significant.bed lists merged regions from CSAW with FDR lower than the threshold provided by the user (default 0.05).
- Looking at 'unchanged' regions is tricky, as there is no statistical test that specifically tests for lack of change, and the test for differential binding is not symmetric. Regions that are not passing the FDR threshold will be a mix  of trully unchanged regions as well of trully changed regions for which you didn't have enough power to detect differential binding at given FDR. Perhaps looking at regions with good correlation across groups would be a workaround. Or regions with very small log-fold changes between groups.
- Filtered.results.bed is additionally filtered on logFC. If you don't see a difference between DiffBinding_allregions.bed and Filtered.results.bed, this suggests all the regions in DiffBinding_allregions.bed passed the FDR and LFC thresholds used to produce Filtered.results.bed.  The filtering is done the same way between ATAC-seq and ChIP-seq workflows. 

- DiffBinding_significant.bed is written with rtracklayer::export.bed, while Filtered.results.bed with write.table. 
The coordinates in the first one are correct (0-based). I'll add a fix for the coordinates in the latter table.

Best,

Katarzyna

**Comment by katsikora** [comment]
Coordinates in Filtered.results.bed are now fixed in release 2.2.3.
